lib/ddtrace/sampler.rb

```diff
@@ -229,8 +237,33 @@ def priority_assigned_upstream?(span)
       span.context && !span.context.sampling_priority.nil?
     end
 
-    def priority_sample(span)
-      @priority_sampler.sample?(span)
+    def priority_sample!(span)
+      preserving_sampling(span) do
+        @priority_sampler.sample!(span)
+      end
+    end
+
+    # Ensures the span is always propagated to the writer and that
+    # the sample rate metric represents the true client-side sampling.
+    def preserving_sampling(span)
+      pre_sample_rate_metric = span.get_metric(SAMPLE_RATE_METRIC_KEY)
+
+      yield.tap do
+        # NOTE: We'll want to leave `span.sampled = true` here; all spans for priority sampling must
+        #       be sent to the agent. Otherwise metrics for traces will not be accurate, since the
+        #       agent will have an incomplete dataset.
+        #
+        #       We also ensure that the agent knows we that our `post_sampler` is not performing true sampling,
+        #       to avoid erroneous metric upscaling.
+        span.sampled = true
+        if pre_sample_rate_metric
+          # Restore true sampling metric, as only the @pre_sampler can reject traces
+          span.set_metric(SAMPLE_RATE_METRIC_KEY, pre_sample_rate_metric)
```

Comment:
This shouldn't be unset by any sampler right?

Reply:
It also shouldn't be set by any other sampler either

Reply:
This is only needed for samplers which cause traces to not be send to the agent `span.sampled = false`

Reply:
The default "post_sampler", `RateByServiceSampler`: https://github.com/DataDog/dd-trace-rb/blob/94b47a66ce57ff1e692920c320fc8040b3e63366/lib/ddtrace/sampler.rb#L192
ultimately delegate the sampling decision to an instance of `RateSampler`. That sampler sets this metric whenever it successfully samples: https://github.com/DataDog/dd-trace-rb/blob/94b47a66ce57ff1e692920c320fc8040b3e63366/lib/ddtrace/sampler.rb#L71-L73

In case case we restore it to the true sampling metric value, store under `pre_sample_rate_metric`.
No sampler today is unsetting this metric, but code is put in place for completeness.

#### Backstory
Ultimately, the current contract interface we have for samplers (`sample?` for no side-effects, and `sample!` for true sampling decision with side-effects) does not correctly allow the samplers to be used as both: 
1. Top-level sampler: which ultimately makes the sampling decision, possibly setting metrics and tags.
2. Chained sampler: provide a decision, but without side-effects. One example is the  "post_sampler" in `PrioritySampler` above.

We have cases where we want "some" side-effects, like in the `RuleSampler` when used in a sampler chain: it needs to set rule-specific metrics (e.g. rule sampling ratio metric), but not actually set the sampling decision. It's a "middle-ground" between the `sample!` and `sample?` of today.
I chose to go with an implementation with reduce changes to the existing samplers because:
* I did not have a clear vision on how to have a clean contract to represent all these requirements I mentioned above in a concise package.
* I did not want to perform a major refactoring on the existing samplers, introduction risk to users of the existing sampler.